[SPARK-35345][SQL] Add Parquet tests to BloomFilterBenchmark

huaxingao opened the pull request:
### What changes were proposed in this pull request?
Add BloomFilter Benchmark test for Parquet


### Why are the changes needed?
Currently, we only have BloomFilter Benchmark test for ORC. Will add one for Parquet too.

### Does this PR introduce _any_ user-facing change?
no


### How was this patch tested?
tested the newly added benchmark test

huaxingao committed:
[SPARK-35345][SQL] Add BloomFilter Benchmark test for Parquet

huaxingao committed:
enable ParquetInputFormat bloom filter to true

huaxingao committed:
fix lint scala

SparkQA commented:
**[Test build #138272 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138272/testReport)** for PR 32473 at commit [`1bb6675`](https://github.com/apache/spark/commit/1bb6675f6e09e9ca844ec021e310316584bdf876).

dongjoon-hyun reviewed (commented):
Thank you, @huaxingao .

dongjoon-hyun reviewed (dismissed):
Could you update the benchmark test result here?
- https://github.com/apache/spark/blob/master/sql/core/benchmarks/BloomFilterBenchmark-results.txt
- https://github.com/apache/spark/blob/master/sql/core/benchmarks/BloomFilterBenchmark-jdk11-results.txt

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42794/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/42794/

SparkQA commented:
**[Test build #138272 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138272/testReport)** for PR 32473 at commit [`1bb6675`](https://github.com/apache/spark/commit/1bb6675f6e09e9ca844ec021e310316584bdf876).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138272/

huaxingao committed:
update benchmark test result

SparkQA commented:
**[Test build #138296 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138296/testReport)** for PR 32473 at commit [`2a3f5eb`](https://github.com/apache/spark/commit/2a3f5eb1e6a698b3627e5692211319fba8762582).

SparkQA commented:
**[Test build #138296 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138296/testReport)** for PR 32473 at commit [`2a3f5eb`](https://github.com/apache/spark/commit/2a3f5eb1e6a698b3627e5692211319fba8762582).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42818/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42818/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/42818/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138296/

dongjoon-hyun commented on sql/core/benchmarks/BloomFilterBenchmark-jdk11-results.txt:
```diff
@@ -2,23 +2,45 @@
+Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
+------------------------------------------------------------------------------------------------------------------------
+Without bloom filter                               1201           1230          41         83.3          12.0       1.0X
+With bloom filter                                  1262           1301          54         79.2          12.6       1.0X
```
It looks like strange to me, too.
> The benchmark test result for Parquet bloom filter is quite depressing. Did I do anything wrong?

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/BloomFilterBenchmark.scala:
```diff
@@ -81,8 +80,51 @@ object BloomFilterBenchmark extends SqlBasedBenchmark {
+        }
+        benchmark.addCase("With bloom filter") { _ =>
+          df.write.mode("overwrite")
+            .option(ParquetOutputFormat.BLOOM_FILTER_ENABLED + "#value", true)
```
Please debug with this code path. This might be insufficient to enable bloom filters in Parquet library for some reason. The usual suspects are 1) the parameter is not handed over correctly to Parquet, 2) the parameter needs another parameters, 3) this specific data type is not supported yet in Parquet.
```scala
.option(ParquetOutputFormat.BLOOM_FILTER_ENABLED + "#value", true)
```

cc @ggershinsky since he is a Parquet committer.

ggershinsky replied:
cc @chenjunjiedada

dongjoon-hyun commented:
BTW, thank you for leading this effort in Apache Spark community, @huaxingao . Since this is the first try, you will remove the roadblocks in advance for the users.

sunchao commented:
cc @wangyum too

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/BloomFilterBenchmark.scala:
```diff
@@ -81,8 +80,51 @@ object BloomFilterBenchmark extends SqlBasedBenchmark {
+    withTempPath { dir =>
+      val path = dir.getCanonicalPath
+
+      df.write.parquet(path + "/withoutBF")
+      df.write.option(ParquetOutputFormat.BLOOM_FILTER_ENABLED + "#value", true)
+        .parquet(path + "/withBF")
```
You need to set row group size, e.g.

```scala
df.write.option(ParquetOutputFormat.BLOOM_FILTER_ENABLED + "#value", true)
  .option("parquet.block.size", 1024 * 1024)
  .parquet(path + "/withBF")
```

Then you will see the benchmark difference.

```
[info] Running benchmark: Read a row from 100M rows
[info]   Running case: Without bloom filter
[info]   Stopped after 3 iterations, 2674 ms
[info]   Running case: With bloom filter
[info]   Stopped after 5 iterations, 2383 ms
[info] OpenJDK 64-Bit Server VM 1.8.0_265-b01 on Mac OS X 10.16
[info] Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
[info] Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] Without bloom filter                                872            892          19        114.7           8.7       1.0X
[info] With bloom filter                                   473            477           3        211.4           4.7       1.8X
```

sunchao replied:
Wow that's a great improvement! I wonder if we should make the row group size a parameter for the benchmark too. There seems to be other related parameters too such as `DEFAULT_MAX_BLOOM_FILTER_BYTES`

huaxingao committed:
set parquet block size

SparkQA commented:
**[Test build #138309 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138309/testReport)** for PR 32473 at commit [`10d7a97`](https://github.com/apache/spark/commit/10d7a977391d659d2060ba596c55d0334754866c).

huaxingao committed:
update benchmark result'

huaxingao commented:
jdk8
```
================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 1.8.0_292-b10 on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 524288            1035           1039           6         96.7          10.3       1.0X
With bloom filter, blocksize: 524288                848            878          26        117.9           8.5       1.2X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 1.8.0_292-b10 on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 1048576            998           1012          21        100.2          10.0       1.0X
With bloom filter, blocksize: 1048576               459            484          29        218.1           4.6       2.2X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 1.8.0_292-b10 on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 2097152            948            961          18        105.5           9.5       1.0X
With bloom filter, blocksize: 2097152               275            288          13        363.3           2.8       3.4X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 1.8.0_292-b10 on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 3145728            953            964          11        104.9           9.5       1.0X
With bloom filter, blocksize: 3145728               223            236          16        448.3           2.2       4.3X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 1.8.0_292-b10 on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 4194304            945            966          18        105.8           9.5       1.0X
With bloom filter, blocksize: 4194304               195            209          15        511.8           2.0       4.8X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 1.8.0_292-b10 on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 5242880            984            996          13        101.7           9.8       1.0X
With bloom filter, blocksize: 5242880               214            224          12        468.1           2.1       4.6X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 1.8.0_292-b10 on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 6291456            978            996          19        102.3           9.8       1.0X
With bloom filter, blocksize: 6291456               281            289          10        355.6           2.8       3.5X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 1.8.0_292-b10 on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 7340032            989           1008          19        101.1           9.9       1.0X
With bloom filter, blocksize: 7340032               337            345           6        296.8           3.4       2.9X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 1.8.0_292-b10 on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 8388608           1005           1011           8         99.5          10.0       1.0X
With bloom filter, blocksize: 8388608               385            404          20        259.6           3.9       2.6X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 1.8.0_292-b10 on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 9437184            992           1002          14        100.8           9.9       1.0X
With bloom filter, blocksize: 9437184               521            538          16        191.9           5.2       1.9X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 1.8.0_292-b10 on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
Read a row from 100M rows:                 Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
-------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 10485760            969            998          27        103.2           9.7       1.0X
With bloom filter, blocksize: 10485760               487            510          16        205.3           4.9       2.0X

```

jdk11
```
================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 11.0.11+9-LTS on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 524288            1214           1328         161         82.4          12.1       1.0X
With bloom filter, blocksize: 524288               1192           1358         235         83.9          11.9       1.0X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 11.0.11+9-LTS on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 1048576           1114           1146          45         89.8          11.1       1.0X
With bloom filter, blocksize: 1048576               634            705         123        157.8           6.3       1.8X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 11.0.11+9-LTS on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 2097152           1071           1099          40         93.4          10.7       1.0X
With bloom filter, blocksize: 2097152               468            484          10        213.6           4.7       2.3X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 11.0.11+9-LTS on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 3145728           1081           1103          32         92.5          10.8       1.0X
With bloom filter, blocksize: 3145728               332            369          44        301.6           3.3       3.3X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 11.0.11+9-LTS on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 4194304           1057           1070          20         94.7          10.6       1.0X
With bloom filter, blocksize: 4194304               295            342          31        339.3           2.9       3.6X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 11.0.11+9-LTS on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 5242880           1006           1049          60         99.4          10.1       1.0X
With bloom filter, blocksize: 5242880               237            278          22        422.5           2.4       4.3X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 11.0.11+9-LTS on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 6291456            988           1040          74        101.2           9.9       1.0X
With bloom filter, blocksize: 6291456               268            305          27        373.6           2.7       3.7X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 11.0.11+9-LTS on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 7340032           1004           1031          38         99.6          10.0       1.0X
With bloom filter, blocksize: 7340032               476            547          60        210.2           4.8       2.1X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 11.0.11+9-LTS on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 8388608           1452           1469          25         68.9          14.5       1.0X
With bloom filter, blocksize: 8388608               658            677          17        152.0           6.6       2.2X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 11.0.11+9-LTS on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
Read a row from 100M rows:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 9437184           1456           1477          30         68.7          14.6       1.0X
With bloom filter, blocksize: 9437184               779            825          40        128.4           7.8       1.9X


================================================================================================
Parquet Read
================================================================================================

OpenJDK 64-Bit Server VM 11.0.11+9-LTS on Linux 5.4.0-1046-azure
Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
Read a row from 100M rows:                 Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
-------------------------------------------------------------------------------------------------------------------------
Without bloom filter, blocksize: 10485760           1472           1474           3         68.0          14.7       1.0X
With bloom filter, blocksize: 10485760               777            830          65        128.7           7.8       1.9X
```

huaxingao commented:
The result looks really good now. Thanks everyone! Your guys are extremely helpful!

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/BloomFilterBenchmark.scala:
```diff
@@ -81,8 +80,57 @@ object BloomFilterBenchmark extends SqlBasedBenchmark {
+      withTempPath { dir =>
+        val path = dir.getCanonicalPath
+
+        df.write.parquet(path + "/withoutBF")
```
When comparing withoutBF and withBF, I think we should use the same `parquet.block.size` for both cases. It might also affects the numbers of withoutBF.

huaxingao replied:
Sorry, forgot to change that. Fixed. Will update results

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42831/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42831/

huaxingao committed:
set parquet.block.size for withoutBF

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/42831/

SparkQA commented:
**[Test build #138311 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138311/testReport)** for PR 32473 at commit [`34d0511`](https://github.com/apache/spark/commit/34d05113d307395bd1c1449651e09a8285fd0c6e).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42833/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42833/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/42833/

dongjoon-hyun reviewed (commented):
Shall we change the grouping in order see the trend according to the block size? For example,

```
...
Without bloom filter, blocksize: 8388608           1005           1011           8         99.5          10.0       1.0X
Without bloom filter, blocksize: 9437184            992           1002          14        100.8           9.9       1.0X
With bloom filter, blocksize: 8388608               385            404          20        259.6           3.9       2.6X
With bloom filter, blocksize: 9437184               521            538          16        191.9           5.2       1.9X
...
```

huaxingao committed:
update results

SparkQA commented:
**[Test build #138312 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138312/testReport)** for PR 32473 at commit [`21cc2ac`](https://github.com/apache/spark/commit/21cc2ac907ffe9256942d818663ce225d1a1b992).

wangyum commented on sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/BloomFilterBenchmark.scala:
```diff
@@ -81,8 +80,57 @@ object BloomFilterBenchmark extends SqlBasedBenchmark {
+      withTempPath { dir =>
+        val path = dir.getCanonicalPath
+
+        df.write.option("parquet.block.size", blocksize).parquet(path + "/withoutBF")
```
Could we use the same value for block size and compression size? Please see how we did it in [FilterPushdownBenchmark](https://github.com/apache/spark/blob/7158e7f986630d4f67fb49a206d408c5f4384991/sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/FilterPushdownBenchmark.scala#L61-L62).

huaxingao replied:
@wangyum Sorry, I am new to parquet. Somehow I didn't see parquet has compression size, seems only ORC has `orc.compress.size`?

SparkQA commented:
**[Test build #138309 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138309/testReport)** for PR 32473 at commit [`10d7a97`](https://github.com/apache/spark/commit/10d7a977391d659d2060ba596c55d0334754866c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42834/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42834/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/42834/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138309/

SparkQA commented:
**[Test build #138311 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138311/testReport)** for PR 32473 at commit [`34d0511`](https://github.com/apache/spark/commit/34d05113d307395bd1c1449651e09a8285fd0c6e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138311/

SparkQA commented:
**[Test build #138312 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138312/testReport)** for PR 32473 at commit [`21cc2ac`](https://github.com/apache/spark/commit/21cc2ac907ffe9256942d818663ce225d1a1b992).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138312/

ggershinsky commented:
> I wonder if we should make the row group size a parameter for the benchmark too. 

> Shall we change the grouping in order see the trend according to the block size? For example,
> 
> ```
> ...
> Without bloom filter, blocksize: 8388608           1005           1011           8         99.5          10.0       1.0X
> Without bloom filter, blocksize: 9437184            992           1002          14        100.8           9.9       1.0X
> With bloom filter, blocksize: 8388608               385            404          20        259.6           3.9       2.6X
> With bloom filter, blocksize: 9437184               521            538          16        191.9           5.2       1.9X
> ...
> ```


+1 to these proposals. Bloom filtering is done per row group (column chunk), so I guess having smaller row groups makes it more efficient. The default parquet.block.size is 128MB. It would be interesting to see a table with block size values ranging from 1MB to 128MB (or maybe higher), eg in multiples of 2, like 1, 2, 4, 8, 16, 32, 64, 128 MB.

ggershinsky commented:
Thanks @huaxingao . I think my basic question is about the extent to which these results are representative for a typical user. If the default block size is 128MB, providing numbers for 0.5-10MB blocks might seem to be unhelpful. If the recommendation is to use 1 MB or 4MB block size, this is a problem, because the default page size in Parquet is 1MB; having a very small block with one or a few pages might be good for bloom filtering, but is bad for other performance optimizations. The 128MB blocks could be bad for bloom filtering because there is just one row group when you write 100M records. And there are other parameters, like the `DEFAULT_MAX_BLOOM_FILTER_BYTES` mentioned by @sunchao . So this is not a comprehensive benchmark.
But maybe it's ok. Still, at a minimum, I would recommend removing the results for 0.5MB and 1MB block sizes- because row groups that small, with just one data page, don't make much sense. I would also suggest adding measurements for 16, 64 and 128MB; this might be of some help for the users of parquet bloom filters.

huaxingao committed:
adding measurements for 16, 64 and 128M

SparkQA commented:
**[Test build #138410 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138410/testReport)** for PR 32473 at commit [`47f70b7`](https://github.com/apache/spark/commit/47f70b7fa00512e9c971c74ba3714062c7fb216e).

huaxingao committed:
update test result

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42932/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42932/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/42932/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42933/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42933/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/42933/

SparkQA commented:
**[Test build #138410 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138410/testReport)** for PR 32473 at commit [`47f70b7`](https://github.com/apache/spark/commit/47f70b7fa00512e9c971c74ba3714062c7fb216e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138410/

SparkQA commented:
**[Test build #138411 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138411/testReport)** for PR 32473 at commit [`05b20af`](https://github.com/apache/spark/commit/05b20af9443b653841a6497eb52aef40fb6058e3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138411/

xinrong-meng committed:
[SPARK-35098][PYTHON] Re-enable pandas-on-Spark test cases

Re-enable some pandas-on-Spark test cases.

pandas version in GitHub Actions is upgraded now so we can re-enable  some pandas-on-Spark test cases.

No.

Unit tests.

Closes #32682 from xinrong-databricks/enable_tests.

Authored-by: Xinrong Meng <xinrong.meng@databricks.com>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>

huaxingao committed:
add benchmark test for IN set with 3000 predicates

SparkQA commented:
**[Test build #139011 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139011/testReport)** for PR 32473 at commit [`df044b3`](https://github.com/apache/spark/commit/df044b3d76e5be496d86ad0f47e91cf6bb20becb).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43529/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43529/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43529/

gengliangwang committed:
[SPARK-35535][SQL] New data source V2 API: LocalScan

### What changes were proposed in this pull request?

Add a new data source V2 API: `LocalScan`. It is a special Scan that will happen on Driver locally instead of Executors.

### Why are the changes needed?

The new API improves the flexibility of the DSV2 API. It allows developers to implement connectors for data sources of small data sizes.
For example, we can build a data source for Spark History applications from Spark History Server RESTFUL API. The result set is small and fetching all the results from the Spark driver is good enough. Making it a data source allows us to operate SQL queries with filters or table joins.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test

Closes #32678 from gengliangwang/LocalScan.

Lead-authored-by: Gengliang Wang <ltnwgl@gmail.com>
Co-authored-by: Gengliang Wang <gengliang@apache.org>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>
(cherry picked from commit 5bcd1c29f0e8442d73d3985253ccadb6f4078044)

SparkQA commented:
**[Test build #139011 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139011/testReport)** for PR 32473 at commit [`df044b3`](https://github.com/apache/spark/commit/df044b3d76e5be496d86ad0f47e91cf6bb20becb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139011/

huaxingao committed:
change the num of predicates in IN set to 300 for now

SparkQA commented:
**[Test build #139030 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139030/testReport)** for PR 32473 at commit [`e582c26`](https://github.com/apache/spark/commit/e582c26661bd2258bdf18355e50e3bc204867402).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43548/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43548/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43548/

SparkQA commented:
**[Test build #139030 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139030/testReport)** for PR 32473 at commit [`e582c26`](https://github.com/apache/spark/commit/e582c26661bd2258bdf18355e50e3bc204867402).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139030/

huaxingao committed:
comment out parquet IN set test for now

SparkQA commented:
**[Test build #139045 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139045/testReport)** for PR 32473 at commit [`d4d39d3`](https://github.com/apache/spark/commit/d4d39d3fdecccd3551b07f8249a4015a0420a170).

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43565/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43565/

SparkQA commented:
**[Test build #139045 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139045/testReport)** for PR 32473 at commit [`d4d39d3`](https://github.com/apache/spark/commit/d4d39d3fdecccd3551b07f8249a4015a0420a170).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139045/

karenfeng committed:
[SPARK-35194][SQL] Refactor nested column aliasing for readability

### What changes were proposed in this pull request?

Refactors `NestedColumnAliasing` and `GeneratorNestedColumnAliasing` for readability.

### Why are the changes needed?

Improves readability for future maintenance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing tests.

Closes #32301 from karenfeng/refactor-nested-column-aliasing.

Authored-by: Karen Feng <karen.feng@databricks.com>
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

sarutak committed:
[SPARK-35194][SQL][FOLLOWUP] Recover build error with Scala 2.13 on GA

### What changes were proposed in this pull request?

This PR fixes a build error with Scala 2.13 on GA.
#32301 seems to bring this error.

### Why are the changes needed?

To recover CI.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

GA

Closes #32696 from sarutak/followup-SPARK-35194.

Authored-by: Kousuke Saruta <sarutak@oss.nttdata.com>
Signed-off-by: Kousuke Saruta <sarutak@oss.nttdata.com>

cloud-fan committed:
[SPARK-35559][TEST] Speed up one test in AdaptiveQueryExecSuite

### What changes were proposed in this pull request?

I just noticed that `AdaptiveQueryExecSuite.SPARK-34091: Batch shuffle fetch in AQE partition coalescing` takes more than 10 minutes to finish, which is unacceptable.

This PR sets the shuffle partitions to 10 in that test, so that the test can finish with 5 seconds.

### Why are the changes needed?

speed up the test

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

N/A

Closes #32695 from cloud-fan/test.

Authored-by: Wenchen Fan <wenchen@databricks.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

huaxingao committed:
change num of predicate to 30

SparkQA commented:
**[Test build #139074 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139074/testReport)** for PR 32473 at commit [`c0e9d97`](https://github.com/apache/spark/commit/c0e9d97199a88a028133a79d05563b7001255e67).

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43595/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43595/

SparkQA commented:
**[Test build #139074 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139074/testReport)** for PR 32473 at commit [`c0e9d97`](https://github.com/apache/spark/commit/c0e9d97199a88a028133a79d05563b7001255e67).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139074/

huaxingao committed:
change # of predicate

SparkQA commented:
**[Test build #139517 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139517/testReport)** for PR 32473 at commit [`89eb722`](https://github.com/apache/spark/commit/89eb7229bb050fae5d76d1173e54fdff113678db).

huaxingao committed:
Revert "change # of predicate"

This reverts commit 89eb7229bb050fae5d76d1173e54fdff113678db.

huaxingao committed:
Revert "[SPARK-35098][PYTHON] Re-enable pandas-on-Spark test cases"

This reverts commit b9e4faf2ed6b1d7e79af07ae3409e9a50f6844c2.

huaxingao committed:
Revert "[SPARK-35194][SQL][FOLLOWUP] Recover build error with Scala 2.13 on GA"

This reverts commit 0a2edadedfbcae8b9ed6ff4f04ec9cd39bf50ba7.

huaxingao committed:
Revert "[SPARK-35194][SQL] Refactor nested column aliasing for readability"

This reverts commit cf8e6d63a05b314c45f0d427de3f6889d99e8d3c.

huaxingao committed:
Revert "[SPARK-35535][SQL] New data source V2 API: LocalScan"

This reverts commit e5ee38c71fcdca68838caa1e672ec06f5de5fc26.

huaxingao committed:
Revert "[SPARK-35559][TEST] Speed up one test in AdaptiveQueryExecSuite"

This reverts commit 5fc105f70a50160954b593ac2c1da3010310255e.

SparkQA commented:
**[Test build #139523 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139523/testReport)** for PR 32473 at commit [`324fedb`](https://github.com/apache/spark/commit/324fedb359fe05c97c75db641fabcb5ea957f0a0).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44042/

huaxingao committed:
reduce # of predicate

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44042/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44042/

SparkQA commented:
**[Test build #139527 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139527/testReport)** for PR 32473 at commit [`d6e320c`](https://github.com/apache/spark/commit/d6e320c5ad1ea3d7f248525bf9e333875c0887f7).

huaxingao committed:
update test results

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44048/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44048/

SparkQA commented:
**[Test build #139533 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139533/testReport)** for PR 32473 at commit [`d481ec1`](https://github.com/apache/spark/commit/d481ec1f0a874d1b42521111f9b8c37549cb0586).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44052/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44052/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44058/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44052/

viirya reviewed (commented):
Is the benchmark run with the built-in In predicate in Parquet? Or still using current one in Spark?

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44058/

SparkQA commented:
**[Test build #139523 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139523/testReport)** for PR 32473 at commit [`324fedb`](https://github.com/apache/spark/commit/324fedb359fe05c97c75db641fabcb5ea957f0a0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44058/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139523/

SparkQA commented:
**[Test build #139527 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139527/testReport)** for PR 32473 at commit [`d6e320c`](https://github.com/apache/spark/commit/d6e320c5ad1ea3d7f248525bf9e333875c0887f7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139527/

SparkQA commented:
**[Test build #139533 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139533/testReport)** for PR 32473 at commit [`d481ec1`](https://github.com/apache/spark/commit/d481ec1f0a874d1b42521111f9b8c37549cb0586).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139533/

SparkQA commented:
**[Test build #139517 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139517/testReport)** for PR 32473 at commit [`89eb722`](https://github.com/apache/spark/commit/89eb7229bb050fae5d76d1173e54fdff113678db).
 * This patch **fails from timeout after a configured wait of `500m`**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139517/

huaxingao commented:
@viirya 
> Is the benchmark run with the built-in In predicate in Parquet? Or still using current one in Spark?

It's still using the current one in Spark. That's why I kept on decreasing the num of predicates, otherwise, i got OOM.

viirya commented on sql/core/benchmarks/BloomFilterBenchmark-jdk11-results.txt:
```diff
@@ -2,23 +2,179 @@
+Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz
+Read a row from 1M rows:                  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
+------------------------------------------------------------------------------------------------------------------------
+Without bloom filter                                 70             76           6         14.2          70.2       1.0X
+With bloom filter                                    73            103          22         13.8          72.6       1.0X
```
For JDK8, bloom filter seems faster a bit.

huaxingao replied:
Not due to IN predicate problem because ORC also seems a bit slower with bloom filter. I think the data is too small. Let me increase the data size and try again.

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/BloomFilterBenchmark.scala:
```diff
@@ -81,8 +85,113 @@ object BloomFilterBenchmark extends SqlBasedBenchmark {
+      df2.repartition(col("value")).sort(col("value"))
+        .write.option("orc.bloom.filter.columns", "value").orc(path + "/withBF")
+
+      runBenchmark(s"ORC Read for IN set") {
```
nit: s"" is not needed.

viirya reviewed (commented):
Some unnecessary s"". And a question about IN predicate with bloom filter case in Parquet. Otherwise looks okay.

huaxingao committed:
address comment

huaxingao committed:
fix error

SparkQA commented:
**[Test build #139689 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139689/testReport)** for PR 32473 at commit [`82e1e8e`](https://github.com/apache/spark/commit/82e1e8ed22b968e846c44fe6eb1c48a5cd219822).

SparkQA commented:
**[Test build #139689 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139689/testReport)** for PR 32473 at commit [`82e1e8e`](https://github.com/apache/spark/commit/82e1e8ed22b968e846c44fe6eb1c48a5cd219822).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #142484 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142484/testReport)** for PR 32473 at commit [`82e1e8e`](https://github.com/apache/spark/commit/82e1e8ed22b968e846c44fe6eb1c48a5cd219822).

SparkQA commented:
**[Test build #142490 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142490/testReport)** for PR 32473 at commit [`82e1e8e`](https://github.com/apache/spark/commit/82e1e8ed22b968e846c44fe6eb1c48a5cd219822).

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46995/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46995/

SparkQA commented:
**[Test build #142484 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142484/testReport)** for PR 32473 at commit [`82e1e8e`](https://github.com/apache/spark/commit/82e1e8ed22b968e846c44fe6eb1c48a5cd219822).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #142490 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142490/testReport)** for PR 32473 at commit [`82e1e8e`](https://github.com/apache/spark/commit/82e1e8ed22b968e846c44fe6eb1c48a5cd219822).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/142490/

huaxingao commented:
This PR is kind of messy. I will close this. The new PR is here https://github.com/apache/spark/pull/34594
